standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1740,4039 +649,413 @@ public long getLatestTxnIdInConflict(long txnid) throws MetaException {
+   * mode and response gets lost.  Then {@link org.apache.hadoop.hive.metastore.RetryingMetaStoreClient}
+   * retries, and enqueues another set of locks in LOCK_WAITING.  The 2nd LockResponse is delivered
+   * to the DbLockManager, which will keep dong {@link #checkLock(CheckLockRequest)} until the 1st
+   * set of locks times out.
    */
-  void checkRetryable(SQLException e, String caller, boolean retryOnDuplicateKey)
-      throws RetryException {
-
-    // If you change this function, remove the @Ignore from TestTxnHandler.deadlockIsDetected()
-    // to test these changes.
-    // MySQL and MSSQL use 40001 as the state code for rollback.  Postgres uses 40001 and 40P01.
-    // Oracle seems to return different SQLStates and messages each time,
-    // so I've tried to capture the different error messages (there appear to be fewer different
-    // error messages than SQL states).
-    // Derby and newer MySQL driver use the new SQLTransactionRollbackException
-    boolean sendRetrySignal = false;
+  @RetrySemantics.CannotRetry
+  public LockResponse lock(LockRequest rqst) throws NoSuchTxnException, TxnAbortedException, MetaException {
     try {
-      if(dbProduct == null) {
-        throw new IllegalStateException("DB Type not determined yet.");
-      }
-      if (dbProduct.isDeadlock(e)) {
-        if (deadlockCnt++ < ALLOWED_REPEATED_DEADLOCKS) {
-          long waitInterval = deadlockRetryInterval * deadlockCnt;
-          LOG.warn("Deadlock detected in {}. Will wait {} ms try again up to {} times.", caller, waitInterval,
-              (ALLOWED_REPEATED_DEADLOCKS - deadlockCnt + 1));
-          // Pause for a just a bit for retrying to avoid immediately jumping back into the deadlock.
-          try {
-            Thread.sleep(waitInterval);
-          } catch (InterruptedException ie) {
-            // NOP
-          }
-          sendRetrySignal = true;
-        } else {
-          LOG.error("Too many repeated deadlocks in {}, giving up.", caller);
-        }
-      } else if (isRetryable(conf, e)) {
-        //in MSSQL this means Communication Link Failure
-        sendRetrySignal = waitForRetry(caller, e.getMessage());
-      } else if (retryOnDuplicateKey && isDuplicateKeyError(e)) {
-        sendRetrySignal = waitForRetry(caller, e.getMessage());
-      }
-      else {
-        //make sure we know we saw an error that we don't recognize
-        LOG.info("Non-retryable error in {} : {}", caller, getMessage(e));
-      }
-    }
-    finally {
-      /*if this method ends with anything except a retry signal, the caller should fail the operation
-      and propagate the error up to the its caller (Metastore client); thus must reset retry counters*/
-      if(!sendRetrySignal) {
-        deadlockCnt = 0;
-        retryNum = 0;
-      }
-    }
-    if(sendRetrySignal) {
-      throw new RetryException();
+      return txnLockManager.checkLock(txnLockManager.enqueueLock(rqst), rqst.getTxnid(), rqst.isZeroWaitReadEnabled(),
```

Comment:
why don't we create `lock` method in txnLockManager?

Reply:
`enqueuelock` also has `SqlRetry` and `Transactional` annotations so it needs to be on the interface and called through the proxy.

Reply:
ok, but can we create a wrapper `lock` method without any retry for simplicity?